pkg/splunk/enterprise/upgrade.go

```diff
@@ -0,0 +1,315 @@
+//  2. Cluster Manager - if LM ref is defined, wait for License manager to complete
+//  3. Monitoring Console - if CM ref is defined, wait for Cluster Manager to complete
+//  4. Search Head Cluster - if MC ref , CM ref , LM ref is defined, wait for them to complete in order,
+//     if any one of them not defined, ignore them and wait for the one added in ref
+//  5. Indexer Cluster - same as above also wait for search head cluster to complete before starting upgrade
+//     if its multisite then do 1 site at a time
+//     function returns bool and error , true  - go ahead with upgrade
+//     false -  exit the reconciliation loop with error
+func UpgradePathValidation(ctx context.Context, c splcommon.ControllerClient, cr splcommon.MetaObject, spec enterpriseApi.CommonSplunkSpec, mgr *indexerClusterPodManager) (bool, error) {
+	reqLogger := log.FromContext(ctx)
+	scopedLog := reqLogger.WithName("isClusterManagerReadyForUpgrade").WithValues("name", cr.GetName(), "namespace", cr.GetNamespace())
+	eventPublisher, _ := newK8EventPublisher(c, cr)
+	kind := cr.GroupVersionKind().Kind
+	scopedLog.Info("kind is set to ", "kind", kind)
+	// start from standalone first
+	goto Standalone
+
+	// if custom resource type is standalone or license manager go ahead and upgrade
+Standalone:
+	if cr.GroupVersionKind().Kind == "Standalone" {
+		return true, nil
+	} else {
+		goto LicenseManager
+	}
+LicenseManager:
+	if cr.GroupVersionKind().Kind == "LicenseManager" {
+		return true, nil
+	} else {
+		licenseManagerRef := spec.LicenseManagerRef
+		// if custom resource type not license manager or standalone then
+		// check if there is license manager reference
+		// if no reference go to cluster manager
+		if licenseManagerRef.Name == "" {
+			goto ClusterManager
+		}
+
+		namespacedName := types.NamespacedName{Namespace: cr.GetNamespace(), Name: licenseManagerRef.Name}
+		licenseManager := &enterpriseApi.LicenseManager{}
+		// get the license manager referred in CR
+		err := c.Get(ctx, namespacedName, licenseManager)
+		if err != nil {
+			if k8serrors.IsNotFound(err) {
+				goto ClusterManager
+			}
+			return false, err
+		}
+
+		// get current image of license manager
+		lmImage, err := getCurrentImage(ctx, c, licenseManager, SplunkLicenseManager)
+		if err != nil {
+			eventPublisher.Warning(ctx, "isClusterManagerReadyForUpgrade", fmt.Sprintf("Could not get the License Manager Image. Reason %v", err))
+			scopedLog.Error(err, "Unable to get licenseManager current image")
+			return false, err
+		}
+		// if license manager status is ready and CR spec and current license manager image are not same
+		// then return with error
+		if licenseManager.Status.Phase != enterpriseApi.PhaseReady || lmImage != spec.Image {
+			return false, fmt.Errorf("license manager is not ready or license manager current image is different than CR image")
+		}
+		goto ClusterManager
+	}
+ClusterManager:
+	if cr.GroupVersionKind().Kind == "ClusterManager" {
```

Comment:
Is it possible to make the if/else under each goto a function and generalize it? If its too much work we can take it as a backlog item. Also some functionality might be different based on the CR.

Reply:
I thought about it, but i felt it makes it easy if I keep everything here. we can plan for that in next cycle.